config.go

```diff
@@ -33,77 +55,74 @@ func (c *Config) setPeriodsAndDates() error {
-			} else {
-				c.startDates = append(c.startDates, date)
-				c.endDates = append(c.endDates, getEndDates(date, frequency.DAILY))
-			}
+	period, err := GetPeriodDifference(startDate, endDate, c.Frequency)
+	if err != nil {
+		return err
+	}
+	c.periods = int64(period)
+	for i := 0; i < period; i++ {
+		date, err := getStartDate(startDate, c.Frequency, i)
+		if err != nil {
+			return err
 		}
-	case frequency.WEEKLY:
-		if days%7 != 0 {
-			return errors.New("uneven end date")
+		if i == 0 {
+			c.startDates = append(c.startDates, c.StartDate)
+		} else {
+			c.startDates = append(c.startDates, *date)
 		}
-		c.periods = days / 7
-		for i := 0; i < int(c.periods); i++ {
-			date := startDate.AddDate(0, 0, 7*i)
-			if i == 0 {
-				c.startDates = append(c.startDates, c.StartDate)
-				c.endDates = append(c.endDates, getEndDates(date, frequency.WEEKLY))
-			} else {
-				c.startDates = append(c.startDates, date)
-				c.endDates = append(c.endDates, getEndDates(date, frequency.WEEKLY))
-			}
-
+		if endDate, err := getEndDates(*date, c.Frequency); err != nil {
+			return err
+		} else {
+			c.endDates = append(c.endDates, endDate)
 		}
+	}
+	return nil
+}
 
+func GetPeriodDifference(from time.Time, to time.Time, freq frequency.Type) (int, error) {
+	var periods int
+	switch freq {
+	case frequency.DAILY:
+		periods = int(to.Sub(from).Hours()/24) + 1
+	case frequency.WEEKLY:
+		days := int(to.Sub(from).Hours()/24) + 1
+		if days%7 != 0 {
+			return -1, ErrUnevenEndDate
+		}
+		periods = days / 7
 	case frequency.MONTHLY:
-		months, err := getMonthsBetweenDates(c.StartDate, c.EndDate)
+		months, err := getMonthsBetweenDates(from, to)
 		if err != nil {
-			return err
+			return -1, err
 		}
-		c.periods = int64(*months)
-		for i := 0; i < int(c.periods); i++ {
-			date := startDate.AddDate(0, i, 0)
-			if i == 0 {
-				c.startDates = append(c.startDates, c.StartDate)
-				c.endDates = append(c.endDates, getEndDates(date, frequency.MONTHLY))
-			} else {
-				c.startDates = append(c.startDates, date)
-				c.endDates = append(c.endDates, getEndDates(date, frequency.MONTHLY))
-			}
-
-		}
-
+		periods = *months
 	case frequency.ANNUALLY:
-		years, err := getYearsBetweenDates(startDate, endDate)
+		years, err := getYearsBetweenDates(from, to)
 		if err != nil {
-			return err
-		}
-		c.periods = int64(*years)
-		for i := 0; i < int(c.periods); i++ {
-			date := startDate.AddDate(i, 0, 0)
-			if i == 0 {
-				c.startDates = append(c.startDates, c.StartDate)
-				c.endDates = append(c.endDates, getEndDates(date, frequency.ANNUALLY))
-			} else {
-				c.startDates = append(c.startDates, date)
-				c.endDates = append(c.endDates, getEndDates(date, frequency.ANNUALLY))
-			}
-
+			return -1, err
 		}
+		periods = *years
 	default:
-		return ErrInvalidFrequency
+		return -1, ErrInvalidFrequency
+	}
+	return periods, nil
+}
 
+func getStartDate(date time.Time, freq frequency.Type, index int) (*time.Time, error) {
```

Comment:
return type doesn't need to be pointer. return zero value of time.Time. 

if you are returning err as last parameter, then you don't need to return pointer for result. caller can check if err is set. if it's set, then something went wrong, and caller has to handle that error. result is expected to be some invalid value in that case.
if error is not set, then result should be a valid value. it can be zero too.

Reply:
got it.